pkgs/development/python-modules/docarray/default.nix

```diff
@@ -0,0 +1,110 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  pythonOlder,
+  poetry-core,
+  numpy,
+  orjson,
+  pydantic,
+  rich,
+  types-requests,
+  typing-inspect,
+  # Optional dependencies
+  av,
+  elasticsearch,
+  elastic-transport,
+  fastapi,
+  hnswlib,
+  jax,
+  lz4,
+  jaxlib,
+  pandas,
+  pillow,
+  types-pillow,
+  protobuf,
+  pymilvus,
+  pydub,
+  qdrant-client,
+  redis,
+  smart-open,
+  torch,
+  trimesh,
+  weaviate-client,
+}:
+buildPythonPackage rec {
+  pname = "docarray";
+  version = "0.40.0";
+  format = "pyproject";
+
+  disabled = pythonOlder "3.8";
+
+  src = fetchFromGitHub {
+    inherit pname version;
+    owner = "docarray";
+    repo = "docarray";
+    rev = "v${version}";
+    hash = "sha256-Vqmfn36UYb3s5Y1OtxtHgU5bCJaWKb5Perqr6X3A9xA=";
+  };
+
+  build-system = [ poetry-core ];
+
+  dependencies = [
+    numpy
+    orjson
+    pydantic
+    rich
+    types-requests
+    typing-inspect
+  ];
+
+  optional-dependencies = {
+    proto = [
+      protobuf
+      lz4
+    ];
+    pandas = [ pandas ];
+    image = [
+      pillow
+      types-pillow
+    ];
+    video = [ av ];
+    audio = [ pydub ];
+    mesh = [ trimesh ];
+    hnswlib = [
+      hnswlib
+      protobuf
+    ];
+    elasticsearch = [
+      elasticsearch
+      elastic-transport
+    ];
+    jac = [
+      # jina-hubble-sdk
+    ];
+    aws = [ smart-open ];
+    torch = [ torch ];
+    web = [ fastapi ];
+    qdrant = [ qdrant-client ];
+    weaviate = [ weaviate-client ];
+    milvus = [ pymilvus ];
+    redis = [ redis ];
+    jax = [
+      jaxlib
+      jax
+    ];
+    epsilla = [
+      # pyepsilla
+    ];
+  };
+
+  pythonImportsCheck = [ "docarray" ];
+
```

Comment:
Could you add `pytestCheckHook` to `nativeCheckInputs`?

Reply:
A lot of test depends on Python packages that are not available yet in `nixpkgs`. Even listing the tests names (to list them in `disabledTests`) is not possible without the dependencies. I can't find a way to disable the tests.

Reply:
I have successfully packaged the missing Python packages locally, i.e. `mktestdocs` and `pyepsilla`. All the test work with them. Should I push them to `nixpkgs` before finishing (and fixing) this PR?

Reply:
You could include both into this PR if possible 👍

Reply:
`pyepsilla` depends on #338831 and on the latest version of `posthog` that should be updated by auto updates.

Reply:
Actually, `mktestdocs` has been added by #338197.